controllers/drain_controller.go

```diff
@@ -0,0 +1,463 @@
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	// create the drain state annotation if it doesn't exist in the sriovNetworkNodeState object
+	nodeStateDrainAnnotationCurrent, NodeStateDrainAnnotationCurrentExist := nodeNetworkState.Annotations[constants.NodeStateDrainAnnotationCurrent]
+	if !NodeStateDrainAnnotationCurrentExist {
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainIdle, dr.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		nodeStateDrainAnnotationCurrent = constants.DrainIdle
+	}
+
+	// create the drain state annotation if it doesn't exist in the node object
+	nodeDrainAnnotation, nodeDrainAnnotationExist := node.Annotations[constants.NodeDrainAnnotation]
+	if !nodeDrainAnnotationExist {
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeDrainAnnotation, constants.DrainIdle, dr.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		nodeDrainAnnotation = constants.DrainIdle
+	}
+
+	//TODO: change this to save it on runtime
+	mcpPauseDrainNodeList := &corev1.NodeList{}
+	err = dr.List(ctx, mcpPauseDrainNodeList, &client.ListOptions{LabelSelector: dr.mcpPauseDrainSelector})
+	if err != nil {
+		// Failed to get node list
+		reqLogger.Error(err, "Error occurred on LIST nodes request from API server")
+		return reconcile.Result{}, err
+	}
+
+	DrainingNodeList := &corev1.NodeList{}
+	err = dr.List(ctx, DrainingNodeList, &client.ListOptions{LabelSelector: dr.mcpPauseDrainSelector})
+	if err != nil {
+		// Failed to get node list
+		reqLogger.Error(err, "Error occurred on LIST nodes request from API server")
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("Max node allowed to be draining at the same time", "MaxParallelNodeConfiguration", maxParallelNodeConfiguration)
+	drainingNodes := len(mcpPauseDrainNodeList.Items) + len(DrainingNodeList.Items)
+	reqLogger.Info("Count of draining", "drainingNodes", drainingNodes)
+
+	// if both are Idle we don't need to do anything
+	if nodeDrainAnnotation == constants.DrainIdle &&
+		nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+		reqLogger.Info("node and nodeState are on idle nothing todo")
+		return reconcile.Result{}, nil
+	}
+
+	if nodeDrainAnnotation == constants.DrainIdle &&
+		nodeStateDrainAnnotationCurrent == constants.DrainComplete {
+		err = dr.completeDrain(ctx, node)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainIdle, dr.Client)
+		return ctrl.Result{}, err
+	}
+
+	if drainingNodes >= maxParallelNodeConfiguration {
+		reqLogger.Info("MaxParallelNodeConfiguration limit reached for draining nodes re-enqueue the request")
+		// TODO: make this time configurable
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	// the node request to drain, but we are at the limit so we re-enqueue the request
```

Comment:
nit: this comment should go a few lines up

Reply:
right!